### PR TITLE
i2c: RTIO driver for stm32

### PR DESCRIPTION
**Pull request by teburd**
Very much a WIP and draft until complete driver for stm32 (targetting v2 to start) using RTIO.

Some ad-hoc commentary here... I'm finding myself copying a lot of logic from i2c_ll_stm32.c and i2c_ll_stm32_v2.c, would it be better to move common code to something like an i2c_ll_stm32_common.c and refactor the existing drivers for this? @erwango your thoughts and opinions would be greatly appreciated on that.

I'm not working towards supporting recovery or i2c target mode just yet but I'm certainly seeing some recurring patterns and would like to perhaps move those into i2c_common.c or i2c_rtio.c to avoid duplicating the work needed to support recovery for sure.

Target mode is really interesting and I'd love to support it, if only for testing, but that will require a bit more thought.

### CODE REVIEW

**Comment by erwango** [comment]
> Some ad-hoc commentary here... I'm finding myself copying a lot of logic from i2c_ll_stm32.c and i2c_ll_stm32_v2.c, would it be better to move common code to something like an i2c_ll_stm32_common.c and refactor the existing drivers for this? @erwango your thoughts and opinions would be greatly appreciated on that.

I see no issue in refactoring the existing driver to avoid duplication and ease maintenance.

**Comment by teburd** [comment]
A first pass review might be worthwhile at this point, I still haven't done a pass on removing code duplication between the drivers here but.... the driver passes my tests now

```
*** Booting Zephyr OS build v3.6.0-940-ge771e9823fa7 ***
Running TESTSUITE i2c_ram
===================================================================
START - test_ram_rtio
submitting write from thread 0x240002b8 addr 7
 PASS - test_ram_rtio in 0.006 seconds
===================================================================
START - test_ram_rtio_isr
timer submitting write, addr e
timer checking write result, submitting read
read command complete
read data complete
 PASS - test_ram_rtio_isr in 0.011 seconds
===================================================================
START - test_ram_transfer
ram using i2c_transfer from thread 0x240002b8 addr 15
 PASS - test_ram_transfer in 0.007 seconds
===================================================================
START - test_ram_write_read
ram using i2c_write and i2c_write_read from thread 0x240002b8 addr 1c
 PASS - test_ram_write_read in 0.008 seconds
===================================================================
TESTSUITE i2c_ram succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [i2c_ram]: pass = 4, fail = 0, skip = 0, total = 4 duration = 0.032 seconds
 - PASS - [i2c_ram.test_ram_rtio] duration = 0.006 seconds
 - PASS - [i2c_ram.test_ram_rtio_isr] duration = 0.011 seconds
 - PASS - [i2c_ram.test_ram_transfer] duration = 0.007 seconds
 - PASS - [i2c_ram.test_ram_write_read] duration = 0.008 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Comment by teburd** [comment]
Adding the autoreload support i'm finding some peculiar behavior that i'm diagnosing, but it makes me believe that autoreload doesn't work like the state diagrams suggest in the stm32h723zg reference manual suggests....

st_i2c_tx

File: drivers/i2c/i2c_ll_stm32_rtio.c [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+static void i2c_stm32_master_mode_end(const struct device *dev)
+{
+	const struct i2c_stm32_config *cfg = dev->config;
+	I2C_TypeDef *i2c = cfg->i2c;
+
+	i2c_stm32_disable_transfer_interrupts(dev);
+
+	if (LL_I2C_IsEnabledReloadMode(i2c)) {
+		LL_I2C_DisableReloadMode(i2c);
+	}
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32_rtio.c:73]: This could be cleaned up if reload is not enabled (or TODOed).

File: drivers/i2c/i2c_ll_stm32_rtio.c [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+
+	/* NACK received */
+	if (LL_I2C_IsActiveFlag_NACK(i2c)) {
+		LL_I2C_ClearFlag_NACK(i2c);
+		/*
+		 * AutoEndMode is always disabled in master mode,
+		 * so send a stop condition manually
+		 */
+		LL_I2C_GenerateStopCondition(i2c);
+		ret = -EIO;
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32_rtio.c:110]: `ret` to be removed

File: drivers/i2c/i2c_ll_stm32_rtio.c (hunk 1 above)
Comment (teburd) [reply drivers/i2c/i2c_ll_stm32_rtio.c:73]: Yes, I plan on dropping reload support to start but it'll be a little bit before I can get back to this PR. Perhaps early next week I'll try poking around again.

**Comment by teburd** [comment]
Dropped for now as autoreload ran into issues on my nucleo h723zg as this was written for whatever reason, likely made a mistake in its implementation or didn't account for something noted in errata/docs somewhere, but moving on for now

**Pull request closed without merging**